### ISSUE TITLE
Limit FastApi to process only 1 Request at a time. 

### ISSUE THREAD

**Issue by arjit1804** [tiangolo/fastapi#4686]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#main.py
from fastapi import FastAPI
app = FastAPI()

@app.post('/limit')
async def func():
    # Doing some work that may take about 10 - 30 minutes
    return result
```

### Description

I want to limit the requests at this endpoint. As this endpoint will take some time to finish, so once a request comes and it starts being processed, any other request that is made before the finishing of first request gets discarded(Or can get 429 response). 

In short  I want to add a Limit of  **1 REQUEST AT A TIME**. 

I've seen slow-api, fastapi-limiter, and other limiters. But all I got from them was to limit rate according to some time like (10 requests/sec) or similar. 

Hope i can get any workaround on this!!

Thanks..

### Operating System

macOS

### FastAPI Version

0.73.0

### Python Version

Python 3.9.10

### Additional Context

<img width="885" alt="image" src="https://user-images.githubusercontent.com/62153451/158328284-767e4e0b-6cee-4ca2-82de-12ea215c68ca.png">
https://nordicapis.com/everything-you-need-to-know-about-api-rate-limiting/

This bottleneck seems to do the trick, but I don't have a clue of implementing this in fastapi.

**Comment by tr-gis** [comment]
did you try locking the threads?.  Seems to work . I tried something like this:

```
#main.py
from fastapi import FastAPI
import threading

lock = threading.Lock()

app = FastAPI()

@app.post('/limit')
def func():
    lock.acquire()
    # Doing some work that may take about 10 - 30 minutes
    lock.release()
    return result
```

Note that I changed function def from async to normal def.

**Comment by iudeen** [comment]
May be this may work:

```
#main.py
from fastapi import FastAPI
app = FastAPI()

def execute(f, *args, **kwargs):
      return f(*args, **kwargs)

@app.post('/limit')
async def func():
    result = execute(long_task())
    # Doing some work that may take about 10 - 30 minutes
    return result

```

**Comment by DanielElisenberg** [comment]
I have encountered a similar problem, where fast-api(starlette?) won't give back a response when multiple clients are waiting their turn to get processed.

Consider an API where we post a lists of words, and our app counts the occurences of each word and persists it in an in-memory dictionary. It then returns the word_counts dictionary to the requester. We can't have more than one request at a time due to race conditions. We can implement this in FastAPI like this:

```py
# app.py
# run this with uvicorn and only one worker:
#     uvicorn app:app

from fastapi import FastAPI
from fastapi.responses import JSONResponse
import uuid
import json


word_counts = {}

app = FastAPI()

@app.post("/")
async def one_at_a_time(words: list):
    req_id = uuid.uuid4()
    print(f"{req_id}: entered")

    for word in words:
        if word in word_counts:
            word_counts[word] += 1
        else:
            word_counts[word] = 1

    print(f"{req_id}: done")
    return JSONResponse(word_counts)
```

If I post 5 requests at the same time:
```py
# clients.py
from multiprocessing import Process
from time import perf_counter
import requests

def send_req():
    words = [f"uniqueword{i}" for i in range(10000000)]
    start = perf_counter()
    requests.post("http://localhost:8000/", json=words)
    print(f"Got response in  {perf_counter()-start}s")

if __name__ == '__main__':
    processes = [Process(target=send_req) for i in range(5)]
    for i in range(5):
        processes[i].start()
        print(f"Started process {i}")
```
This runs synchronously as expected:
```
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
bc9dbf42-88d1-4152-9437-02f0ffb14ac0: entered
bc9dbf42-88d1-4152-9437-02f0ffb14ac0: done
INFO:     127.0.0.1:57516 - "POST / HTTP/1.1" 200 OK
354cfc35-7300-4866-b5ea-e2155db9b235: entered
354cfc35-7300-4866-b5ea-e2155db9b235: done
INFO:     127.0.0.1:57520 - "POST / HTTP/1.1" 200 OK
38465b4a-68cc-4d65-9522-c5c2ef2725a2: entered
38465b4a-68cc-4d65-9522-c5c2ef2725a2: done
INFO:     127.0.0.1:57521 - "POST / HTTP/1.1" 200 OK
aabb06d9-64c3-4c8c-8df5-2655d875b2ea: entered
aabb06d9-64c3-4c8c-8df5-2655d875b2ea: done
INFO:     127.0.0.1:57522 - "POST / HTTP/1.1" 200 OK
27725b2e-337c-4ce3-aba6-0bca87346c5d: entered
27725b2e-337c-4ce3-aba6-0bca87346c5d: done
INFO:     127.0.0.1:57524 - "POST / HTTP/1.1" 200 OK
```
 **but the api does not respond before it is done handling all requests**:
```
Started process 0
Started process 1
Started process 2
Started process 3
Started process 4
Got response in  69.861970134s
Got response in  69.978915942s
Got response in  70.00321651300001s
Got response in  70.03401044399999s
Got response in  70.05243789000001s
```


This is unexpected.
If I recreate a similar api with flask:

```py
# app2.py
# run with gunicorn:
#     gunicorn app2:app

from flask import Flask
from flask import request
import uuid
import json


word_counts = {}

app = Flask(__name__)

@app.route('/', methods=['POST'])
def word_insert():
    words = request.json
    
    req_id = uuid.uuid4()
    print(f"{req_id}: entered")

    for word in words:
        if word in word_counts:
            word_counts[word] += 1
        else:
            word_counts[word] = 1

    print(f"{req_id}: done")
    return json.dumps(word_counts)
```
Making the same requests using the above **clients.py** this runs as I would expect. Synchronously and returning the responses one-by-one:
```
Started process 0
Started process 1
Started process 2
Started process 3
Started process 4
Got response in  17.529333825s
Got response in  32.271863816s
Got response in  46.68059113899999s
Got response in  60.814603995s
Got response in  74.28277882600001s
```
I also recreated the api in starlette, and saw the same unexpected behaviour as with the FastAPI implementation.
As we already use FastAPI within my organization, I would like to use it for synchronous use cases as well. Using "def" rather than "async def" is not a option as the dictionary is no longer threadsafe due to the threadpooling starlette does.

**Comment by itzsimpl** [comment]
@DanielElisenberg have you been able to resolve your issue? Have you contacted starlette?

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

If that solves the original problem, then you can close this issue @arjit1804  ✔️

**Comment by itzsimpl** [comment]
@DanielElisenberg thank you for the update.

**Comment by mironnn** [comment]
> @itzsimpl I figured it out after further investigation: From [starlette discussion](https://github.com/encode/starlette/discussions/1690) discussion:
> 
> > Tried to run the same scripts in a poetry environment, and the bug did not persist.
> > This considering, I suspect the bug occured due to an issue with transative dependencies on my system.
> > Have not been able to reproduce this in controlled poetry environments.
> > Closing this discussion. Feel free to @ me if you also run into a similar problem in the future.
> 
> In other words: There were simply problems with the dependencies on my system

@DanielElisenberg could you please advise, how did you fix the issue?
I'm experiencing the same. FastAPI accepted 10 requests, 1 returned and then 9 returned at the same time.
